design: add initial external authorization proposal

jpeach opened the pull request:
This updates #432.
This updates #2459.

Signed-off-by: James Peach <jpeach@vmware.com>

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,453 @@
+
+## Goals
+- Allow operators to integrate existing authorization services with Contour.
+- Support arbitrary types of authentication and authorization.
```
We should spell out the the specific types this proposal incorporates.

jpeach replied:
This is about creating the mechanism that will let anyone add specific types.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,453 @@
+
+## Non Goals
+
+- Support for multiple external authorization mechanisms.
```
Can you clarify these two bits? I'm not sure what is not going to be a goal.

jpeach replied:
What I meant by this is that different proxies have different protocol mechanisms for incorporating external authorization servers. We aren't going to support multiple mechanisms (i.e. you can't use your external auth server from nginx unmodified).

stevesloka replied:
Sorry still not 100% following you on this thought. Could you rewrite the nongoal in a way that better explains it? Right now it reads as you aren't going to support the auth provider that I want to bring which I know is not the case.

jpeach replied:
> Support for multiple external authorization mechanisms (i.e. the protocol the proxy uses to communicate with the authorization server).

The mechanism is the protocol by which the proxy communicates with the authorization server.

> Various proxies have different mechanisms for integrating external services, but it is impractical to support them all.

Nginx, envoy, traefik, etc all use different external authorization mechanisms (ie. communicating with the authorization server follows different protocols or conventions). What I'm getting at is that it is not a goal for an authorization server written for nginx to work with Contour.

stevesloka replied:
I guess the ask here was to make the doc more clear to remove this confusion.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,453 @@
+Authorization policy can also be set on a `Route`, so that application owners can pass metadata and disable authorization on specific routes.
+
+As discussed in [#2459][3], the Envoy external authorization HTTP filter is configured on the HTTP Connection Manager.
+This means that the scope of how Contour can configure authorizationi depends on the structure of the HTTP configuration.
```
s/authorizationi/authorization

jpeach replied:
Fixed!

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,453 @@
+Each `SupportService` CRD generates a unique upstream Envoy [Cluster][#4], which will emit standard Envoy metrics.
+Note that the Envoy cluster name can be non-obvious, so exposing it in status may be helpful.
+
+The `SupportService` CRD re-uses the `Service` type from the `projectcontour.io/v1` API.
```
I don't think I'm keen on this idea of re-use. If we're going to build a new CRD, let's do that and make the user experience simple.

jpeach replied:
In re-using the types, my goal was to keep it simple, since users should already be familiar with the service struct from using the HTTPProxy.

stevesloka replied:
We got burned by this before when re-using. It's an internal type, I'd rather bite the work to create a 5 line struct vs all the additional logic to make sure users don't reuse this field incorrectly due to the different context.

jpeach replied:
OK, I can copy a matching Service type across, leaving out the unnecessary fields. I guess it will need to have a unique name, since this CRD will eventually migrate to the same package.

youngnick replied:
I think it's clearer to have a similar type that doesn't have restrictions about not setting fields that are present - better to just not have fields that aren't relevant.

jpeach replied:
I looked at this, and we would also have to duplicate the `UpstreamValidation`, `HeadersPolicy` and `LoadBalancerPolicy`. That's possible, but it's also pretty intrusive; it means that we will have to do substantial refactoring to be able to apply helpers like `determineSNI()`, `getProtocol()`, etc. However, it's also possible that this refactoring will end up with cleaner internal APIs (maybe create a `model` package for types that compose the DAG?).

We also need to keep in mind that this API will graduate to `projectcontour/v1` and make sure that what we do now will make sense in terms of that destination. Would we want to deduplicate the types at that point, or keep the more specific types?

youngnick replied:
Really? That much duplication? We can't inline the policies? I only meant that it's better not to use something that has fields that you shouldn't set - that's a pretty confusing UX.

I also hadn't though about methods and type-passing, that will make copying the type a significant refactor. I'm not so enamoured of the idea with that much extra work added.

stevesloka replied:
I posted in slack, but why do you need UpstreamValidation, HeadersPolicy and LoadBalancerPolicy? We're just connecting Envoy to another service. That piece is confusing to me which is why I suggested not to resue the service struct.

jpeach replied:
Envoy should connect to the service over TLS, so you need the `UpstreamValidationPolicy` to express how to check the server certificate. If the service is an externalName service, you might need to rewrite the host header, so you'd need `HeadersPolicy`. There could be other operational reasons to add the headers, e.g. assigning Linkerd headers. The service probably has multiple pods, so `LoadBalancerPolicy` is probably a useful control, otherwise we would just hard-code a load balancing policy.

stevesloka replied:
Those are the bits I don't understand. 

- How would an external name service fit? The goal of this CRD is to connect Envoy to the Auth service over gRPC. Are you saying folks would deploy the Auth server outside the cluster and Envoy would reach out over the external name service? 
- Kubernetes doesn't provide a way to set load balancer policies, is that really required for this implementation?
- Is external name the only reason to have a Headers Policy? This is another field I don't understand how it would be applied in this situation and be useful. 

It seems like the only reason to  have the new crd is to allow for setting conditions, but I have questions about how that will work.

stevesloka replied:
Was your thought on headers policy to map to the different response types like `OkHttpResponse`?   https://www.envoyproxy.io/docs/envoy/v1.14.3/api-v2/service/auth/v2/external_auth.proto#service-auth-v2-deniedhttpresponse

jpeach replied:
> Those are the bits I don't understand.
> 
> * How would an external name service fit? The goal of this CRD is to connect Envoy to the Auth service over gRPC. Are you saying folks would deploy the Auth server outside the cluster and Envoy would reach out over the external name service?

Yes, this would work.

> * Kubernetes doesn't provide a way to set load balancer policies, is that really required for this implementation?

Required? Maybe not. But it's obviously useful and easy to implement.

> * Is external name the only reason to have a Headers Policy? This is another field I don't understand how it would be applied in this situation and be useful.

I gave 2 possible examples above. Probably creative operators can think of more uses for this.

> It seems like the only reason to have the new crd is to allow for setting conditions, but I have questions about how that will work.

OK?

> Was your thought on headers policy to map to the different response types like OkHttpResponse?

No, that's part of the GRPC payload; there's no way to intercept that.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,453 @@
+
+This document does not yet define how Contour monitors the support service.
+However, the presence of the `Conditions` field allows Contour to expose the status of the support service fairly directly.
+At minimum, Contour should monitor the underlying Kubernetes service and expose whether there are sufficient healthy endpoints.
```
Ideally we encourage folks to setup readiness probes and rely on k8s to do the lifting, Contour is just going to record which of those pass the probe. My point here is that Contour shouldn't be doing active checks against the external service.

jpeach replied:
Previously, when dependencies on other services have been proposed, having Contour make active availability checks against those has been suggested. Here, I'm not saying we should actually do that, but it's a possibility that we can leave open.

What I'm suggesting here is consistent with your suggestion I think. Rather than operators checking the state of the underlying service pods, Contour can do that on their behalf (since it already knows about endpoints), and roll that information up to the support service status.

stevesloka replied:
> What I'm suggesting here is consistent with your suggestion I think.

This was not my intent, my thought was to take Contour out of the business of doing these checks. They are external systems that Contour doesn't control, we should use the primitives provided by Kubernetes to implement.

jpeach replied:
> > What I'm suggesting here is consistent with your suggestion I think.
> 
> This was not my intent, my thought was to take Contour out of the business of doing these checks. They are external systems that Contour doesn't control, we should use the primitives provided by Kubernetes to implement.

Yes, I agree with you :)

what I'm trying to say is that "monitor the underlying Kubernetes service" doesn't have to mean "actively health check"; it can mean "see whether there are any endpoints alive", or whatever we agree makes sense. The goal here is that we can take advantage of the CRD status to make the status of the underlying objects more obvious, just like deployments do for pods.

https://github.com/projectcontour/contour/issues/2325

youngnick replied:
Maybe something like `Contour should observe the status of the service, directly or indirectly` would make this clearer.

stevesloka commented on design/external-authorization-design.md:
````diff
@@ -0,0 +1,453 @@
+}
+```
+
+The `Conditions` field in `SupportServicesStatus` follows the standardized upstream structure.
````
Link to upstream reference?

jpeach replied:
I rephrased to reference #2642.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,453 @@
+
+The `AuthorizationServer` field carries and optional default authorization policy.
+This policy sets the default policy for each subsequent `Route`.
+If `authPolicy.Disabled` is true, included `Routes` will not require authorization unless they configure their own authorization policy with `Disabled` set to false.
```
How does `permitInsecure` play into this?

jpeach replied:
That's a good point, I'll check on that.

jpeach replied:
If a route can be insecure, then I don't see a way for the upstream service to guarantee that it can know whether a request has been authorized or not. I updated to specify that insecure routes must disable authorization.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,453 @@
+Since the scope of authorization is quite large, documentation should be written as a new top level section.
+If the documentation is included in the main `httpproxy.md` page, it will be difficult to organize and consume.
+
+### Contour Authserver
```
I think we should just re-use existing things instead of building a new one.

jpeach replied:
I agree. However, we need something to test, which is this. If there's an existing Envoy authserver that covers this case, we would support that.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,453 @@
+  Abstracting the protocol would also work against the goal of being able to integrate existing authorization servers.
+
+## High-Level Design
+A new CRD `SupportService` adds a way to represent and track an authorization service.
```
I think we should just build out a proper CRD for Auth and not try to create a generic one that is useful for many purposes. I feel in the long run, we're going to get into a spot where it's super difficult to setup since it's so generic.

jpeach replied:
I don't think it's tractable to build that. There are an unbounded set of different authorization mechanisms and trying to get them all into a single CRD will be a mess. It has the other disadvantage that it's not extensible by operators, so that only Contour maintainers can implement new authorization types.

stevesloka commented:
Another thought I had was adding auth to `Conditions` somehow, so you could allow operators to tell a child set of Proxies should have specific auth, but then not allow them to configure from there. I think some operators might not want each route to be disabled for example.

jpeach commented:
> Another thought I had was adding auth to `Conditions` somehow, so you could allow operators to tell a child set of Proxies should have specific auth, but then not allow them to configure from there. I think some operators might not want each route to be disabled for example.

I'm not really following this, can you explain this a bit more?

jpeach commented:
> * How does the actual auth flow work?

There's some relevant sequence diagrams in https://github.com/projectcontour/contour/issues/2459. I can reproduce them here for context.

>  You mentioned OIDC, but not how we might configure that?

You would create a `SupportService` CRD that references a service that implements OIDC. That will need to be configured with all the necessary OIDC parameters. There's still a bunch of work needed to explore and document how to do this.

> * You mentioned not allowing external auth servers, so this proposal requires all auth to exist inside the k8s cluster where Contour is running?

If you used an ExternalName Service to back the SupportService CRD, then you could have an auth server outside the cluster. Not sure if that will need special code to support, or whether we would get it for free from Contour's existing Service support.

jpeach commented:
xref #1922

pims commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,453 @@
+## Compatibility
+
+Contour can support v2 of the Envoy GRPC authorization protocol.
+When Contour moves to v3 of the Envoy API, it ccan support both v2 and v3 authorization protocols.
```
nit: `s/ccan/can`

jpeach replied:
Fixed!

pims commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,453 @@
+
+## Implementation
+
+The Contour implementation provides a framework, but isn't useful until authorization services are exist for the protocols that Contour operators need.
```
nit: `s/are//`

jpeach commented:
FYI, @pims convinced me that being able to toggle the fail-open or fail-closed behavior is needed when migrating from internal application auth to external auth, so I'm planning to add that flag to the API.

jpeach commented:
@jpeach Note to self ... consider what happens to headers injected by the external auth server on routes that have auth disabled. A client request could still present those headers, causing an application to become confused (a request could accidentally be performed as an authorized identity).

So perhaps we need to add API to ensure that any injected auth headers are scrubbed from the incoming client request. Envoy has the concept of internal headers, which aligns fairly closely, but that (and the API for removing headers) requires listing specific headers (would be more robust to pattern match IMHO). The Envoy concept on internal requests is also a bit hard to understand and it's not obvious that it would be the right choice in all deployments.

youngnick commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,669 @@
+	// Kind of the referent.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds
+	// +required
+	Kind string `json:"kind,omitempty" protobuf:"bytes,1,opt,name=kind"`
```
Should we make this optional with the behavior being that empty is equivalent to `SupportService`? I know you're basing this off `TypedLocalObjectReference`, but I think it's worth considering a slightly different behavior in this case.

jpeach replied:
If we default the kind, then users would do:
```yaml
    authorization:
      serviceRef:
        apiVersion: projectcontour.io/v1alpha1
        name: htpasswd
        namespace: projectcontour-auth
```

Maybe we could remove or default `apiVersion` as well, depending on how we manage the versioning of this type. I'm also open to just removing these fields altogether.

youngnick replied:
I think it's better to default them than remove them, that gives us the chance to use other kinds later.

skriss replied:
Given that this typed is named `SupportServiceReference`, the `Kind` field seems like it'd be either redundant or contradictory, unless the intent is to allow specializations of SupportService (e.g. AuthSupportService, RateLimitingSupportService)?

youngnick commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,669 @@
+
+	// Context is a set of key/value pairs that are sent to the
+	// authentication server in the check request. If a context
+	// is provided at an enclosing scope, the entries are merges
```
Nit:
```suggestion
	// is provided at an enclosing scope, the entries are merged
```

youngnick commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,669 @@
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+
+	// Context is a set of key/value pairs that are sent to the
```
Do the values need to allow some sort of `%` interpolation or something? That might be worth mentioning if so.

jpeach replied:
There's no interpolation, just arbitrary key/value strings. As we get more implementation experience with authorization servers, some sort of interpolation feature could turn out to be useful.

youngnick replied:
Awesome, I'd prefer no interpolation, that is a whole can of worms. I meant more if it was expected as part of this type of functionality.

youngnick commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,669 @@
+## High-Level Design
+A new `SupportService` CRD adds a way to represent and track an authorization service.
+This CRD is relatively generic, so that it can be re-used for Envoy rate limiting and logging services.
+The core of the `SupportService` CRD is a `projectcontour.v1.HTTPProxy` `Service` specification.
```
This part will need updating if we make a new struct, I would propose something like:
```suggestion
The core of the `SupportService` CRD is a subset of the `projectcontour.v1.HTTPProxy` `Service` specification.
```

youngnick commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,669 @@
+## SupportService status
+
+One of the original requirements for adding dependencies on other services was being able to expose useful operational status.
+The `SupportService` status field is where this information would be exposed, but this design does not address what status should be exposed.
```
I'd say at a minimum there should be a `Ready` condition (in this case indicating that the SupportService is ready to service requests). Probably, we'll want some fields eventually, but `Ready` allows for using the "wait for condition" flow easily.

youngnick replied:
This can be dealt with later though; not blocking.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+    // AuthPolicy updates the authorization policy for client
+	// requests that match this route.
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
```
Is there a way that operators can not allow users to change this? I'm thinking that an operator wants to have auth on for a set of routes, what stops a user from circumventing that and disabling?

jpeach replied:
You'd have to use an admission controller (gatekeeper or suchlike) to control this. Generally, I expect that the app owner is responsible for the app and best placed to tweak authorization.

stevesloka replied:
So this is the same exact use-case of `permitInsecure`. @youngnick I know you had thoughts around this that you don't like the arg today to disable this feature, but I think it needs thought out. 

If admission controller is the answer, we need to scope that work since many folks will want to enforce this from the start.

jpeach replied:
> So this is the same exact use-case of `permitInsecure`. @youngnick I know you had thoughts around this that you don't like the arg today to disable this feature, but I think it needs thought out.
> 
> If admission controller is the answer, we need to scope that work since many folks will want to enforce this from the start.

I'm happy to have this discussion, but I don't think it's practical to have separate admin control over every field that can be specified in a HTTPProxy resource. General purpose admission controllers exist for this reason, and it's hard for me to imagine that doing the equivalent in Contour is really worth it.

youngnick replied:
To be clear, @stevesloka, I don't like that we have to disable a permit globally, that's all. I agree that sometimes, cluster admins are going to want to say "you must have auth on, you can't turn it off", but that conflicts with the goal of being able to selectively disable some routes, which is more useful to start with, I think.

To put this another way, I think that being able to expose *some* parts of an app without auth is more immediately useful than having a way to disable that functionality.

stevesloka replied:
I just think that admins are going to want to put a limitation on this field. It's a big hole in security if teams can purposely or accidentally disable security per route and will want a way to lock this field down to specific roles in the cluster. Just my 2 cents.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+	// authentication check request.
+	//
+	// +optional
+	Authorization *AuthorizationServer `json:"authorization,omitempty"`
```
Should this be part of the `TLS` stuct?

jpeach replied:
It's not TLS configuration, so I didn't put it in that struct.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+}
+
+type SupportServiceStatus struct {
+	Conditions []Condition `json:"conditions"`
```
How do users write to this field? If I build my own auth server to handle my auth mechanism, who manages this field? It's outside the scope of Contour since we may not own the implementation and can't understand what values to write.

jpeach replied:
It's a Contour CRD, so Contour updates the status.

stevesloka replied:
To be clear from another conversation had via Zoom, this statement isn't true. The intent is that both Contour & 3rd party services should update this field.

jpeach replied:
> To be clear from another conversation had via Zoom, this statement isn't true. The intent is that both Contour & 3rd party services should update this field.

Personally, I would be surprised if that happened 🤷

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+    // AuthPolicy updates the authorization policy for client
+	// requests that match this route.
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
```
Can I have an AuthorizationPolicy in an include? I'm thinking of setting a bunch of routes that should be insecure, you could include them and add this to the includes conditions without having to set them on each child route.

jpeach replied:
Does anything else in the HTTPProxy API behaves that way? If there's an established pattern we can try to follow it here, but I don't think that this should be too different from what we already have.

jpeach replied:
We could add a `AuthorizationPolicy` to the include and I can see the usefulness of that. But it's more complexity, so I'd rather wait and address it when there's demand. I'm expecting that `AuthorizationPolicy` won't be tweaked that much in practice.

stevesloka replied:
Just something to think about. By doing this, you'd have an easy way to not need the route disable bit. You could put the insecure routes in an HTTPProxy chain, then the secure in another.

jpeach replied:
Yeh, just a matter of whether we want to do this up front in what will already be a big change.

youngnick replied:
I agree that I think we leave this one and see what our users want. We're not ruling this out by not doing it now.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+	ServiceRef SupportServiceReference `json:"serviceRef"`
+
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
```
Should we not allow `Disabled` at the root level? This essentially would allow users to configure auth, then disable it at the root and child routes would have to enable by setting `Disabled=false` on their route?

jpeach replied:
I think that adding the special case makes it harder for users to reason about. The use case you present here seems like a reasonable one to me; or at least reasonably possible.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+	Context map[string]string `json:"context,omitempty"`
+}
+
+type AuthorizationServer struct {
```
It feels like these settings (AuthPolicy, Timeout, Failopen) should be part of the `SupportService` CRD. That's where the contract to the upstream service is defined / configured. 

If many vhosts utilize the same auth server, you'd have to configure these for each vhost and keep in sync.

jpeach replied:
It can't be part of the CRD. Imagine you have 50 services and 1 auth service. You want to gradually migrate the 50 services over to the centralized auth. This means that you move the services over a few at a time, flipping the fail open policy as you go. This means that these parameters, and the others in this struct are specific to the client.

stevesloka replied:
I think that use-case is only relevant once. The backend auth I doubt would have different timeouts per vhost. If that was the case you'd have different CRDs to represent the different backends. But given many use the same backend auth, you now need to sync those settings manually across many vhosts.

jpeach replied:
Yeh, arguably there is a use case for timeout being on the CRD.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+1. Contour could install itself as the authorization server.
+   This could remove some of the limitations of the Envoy configuration structure at the cost of more complexity in Contour.
+1. Integrate external authorization directly into `HTTPProxy`.
+   This increases the complexity of the `HTTPProxy` structure and makes it difficult to re-use the same authorization service acrtoss multiple proxies.
```
nit: s/acrtoss/across

stevesloka commented:
Also don't want to loose context from a Slack conversation about renaming s/SupportService/ExtensionService: https://kubernetes.slack.com/archives/C8XRH2R4J/p1594707008255100

skriss commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+The drawback of only supporting the GRPC protocol is that many existing Envoy authorization servers only support the HTTP protocol.
+Until Contour adds support for the v3 Envoy API, only the version 2 of the GRPC authorization protocol can be configured.
+
+Operators can bind an authorization service to a `HTTPProxy` using a new field in the `VirtualHost` struct.
```
if I'm understanding correctly, this implies that all routes for a given vhost must use the same auth service (or disable it). Is that workable for users? Are there cases where users would want to use a different auth service for different subsets of the routes under a vhost?

jpeach replied:
It's possible someone would like more flexibility than that, but I've yet to hear from that person. There's no way to specify this at a finer granularity other than funneling everything through an intermediate proxy that we configure.

skriss commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+
+This document does not yet define how Contour monitors the support service.
+However, the presence of the `Conditions` field allows Contour to expose the status of the support service fairly directly.
+At minimum, Contour should observe the underlying Kubernetes Service and expose whether there are sufficient healthy Endpoints.
```
how is "sufficient healthy Endpoints" defined/determined?

jpeach replied:
Probably greater than 0. I expect this to evolve over time.

skriss commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+
+## Implementation
+
+The Contour implementation provides a framework, but isn't useful until authorization services are available for the protocols that Contour operators need.
```
Are we (Contour maintainers) realistically going to be building & maintaining all of the auth services? Or are there other obvious parties who would be interested in collaborating on these or building new ones?

I'm just wondering, if practically we're going to be building and maintaining all of them, if that might change how we'd approach the framework design.

jpeach replied:
I don't know the answers to that. However, it is a requirement that third parties be able to add their own authorization.

jpeach replied:
We should build whatever is needed to make OIDC+OAuth2 work really well.

skriss commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,670 @@
+- Decouple Contour from external authorization, so it can evolve at an independent rate.
+- Integrate cleanly with the HTTPProxy API.
+
+## Non Goals
```
I think it'd be useful to explicitly call out somewhere in Goals or Non-Goals that this design proposes the framework, but does not propose auth server implementations for any specific auth protocols, and that those will also need to be implemented before a user can actually make use of external auth.

jpeach replied:
This topic is discussed later in the doc, but I suppose we could add it here too.

jpeach commented:
I'm pretty sure I addressed all the comments, but I'll leave this until next week to give a change for additional feedback. I added a new "Compatibility" section to make it more obvious how auth will interact with existing features.

stevesloka commented:
> This CRD is relatively generic, so that it can be re-used for Envoy rate limiting and logging services.

If the goal of this CRD is to re-use for other bits other than just auth. How does Contour understand that context? For instance, if we use this for rate limiting and features of the spec aren't valid to be set, how could that be implemented?

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,715 @@
+
+type ExtensionServiceSpec struct {
+	// +required
+	Service contourv1.Service `json:"service"`
```
Just thinking out loud, should `Service` be an array? If you did that like HTTPProxy does, you would be able to shift traffic to new instances of an auth backend via weights.

jpeach replied:
Yeh, I did consider that, but you can already handle this use case (to some extent) with services and labels. If this was a slice, then weights and mirroring could be implemented, which would make it the same as `HTTPProxy`.

stevesloka replied:
> already handle this use case (to some extent) with services and labels

This was the reason HTTPProxy has it as a slice since labels & selectors aren't a great way to effectively implement a rollout.

jpeach commented:
> > This CRD is relatively generic, so that it can be re-used for Envoy rate limiting and logging services.
> 
> If the goal of this CRD is to re-use for other bits other than just auth. How does Contour understand that context?

The context can be understood from where the CRD is referenced. In this design, Contour can do the right thing because the CRD is referenced from an authorization policy. In logging and rate limiting, there would be a similar context in the `HTTPProxy` resource.

>  For instance, if we use this for rate limiting and features of the spec aren't valid to be set, how could that be implemented?

Hard to answer this hypothetical, but since the extension service CRD is basically scoped to generate an Envoy cluster, that's sufficient for the rate limiting and logging cases that we know about. Maybe there could be conflicting use cases in the future, but I expect we could address them by revving the CRD, adding another CRD, or in some other way.

stevesloka commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,715 @@
+- Decouple Contour from external authorization, so it can evolve at an independent rate.
+- Integrate cleanly with the HTTPProxy API.
+
+## Non Goals
```
Also supporting v1/v1beta1.Ingress is not a goal?

jpeach replied:
Yep, so far it is not a goal. In general, when adding features, we haven't also added them to Ingress, so not sure if we want to do that here 🤷

youngnick replied:
I think that we will add these features to HTTPProxy first, and consider if we want to add them to Ingress later. The only way we could do that would be with a forest of annotations, which I am reluctant to do without a good reason.

stevesloka replied:
I'm not suggesting that we add, just to call it out in the design so when someone comes and reads the design doc, it's clear. That was one of the original goals of these designs. It was to plan out how or what a feature looked like, even if parts were far off in the future. 

The second goal was to have these features planned out so that more users could contribute with a plan already in place.

stevesloka commented:
> The context can be understood from where the CRD is referenced. In this design, Contour can do the right thing because the CRD is referenced from an authorization policy. In logging and rate limiting, there would be a similar context in the HTTPProxy resource.

Ahh ok that could work looking at where it's referenced from. =)

jpeach commented:
* added ingress to non-goals
* moved timeout to ExtensionService CRD
* made ExtensionService services field a slice

tsaarni commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,710 @@
+	// protocol options will be available in future.
+	//
+	// +kubebuilder:validation:Enum=v2
+	ProtocolVersion SupportProtocolVersion `json:"protocolVersion,omitempty"`
```
Question: is there timeplan for moving to Envoy API v3?  [Deprecation](https://www.envoyproxy.io/docs/envoy/latest/api/api_supported_versions#api-supported-versions) seems to be scheduled for end of the year.  

Comment: The `transport_api_version` parameters can only be set in [v3](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#extensions-filters-http-ext-authz-v3-extauthz).  So user could only set the `protocolVersion` field after Contour has moved to v3.  I wonder if development of external authorization could be synchronized with the API version move so that v2 support could be skipped completely, and `protocolVersion` field in the spec could be removed?  If the Envoy v2 deprecation period holds, one could set `protocolVersion` to `v2` for only couple of months anyways.

jpeach replied:
> Question: is there timeplan for moving to Envoy API v3?  [Deprecation](https://www.envoyproxy.io/docs/envoy/latest/api/api_supported_versions#api-supported-versions) seems to be scheduled for end of the year.  

There is an issue but it's not being worked on yet.

> 
> Comment: The `transport_api_version` parameters can only be set in [v3](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#extensions-filters-http-ext-authz-v3-extauthz).  So user could only set the `protocolVersion` field after Contour has moved to v3.  I wonder if development of external authorization could be synchronized with the API version move so that v2 support could be skipped completely, and `protocolVersion` field in the spec could be removed?  If the Envoy v2 deprecation period holds, one could set `protocolVersion` to `v2` for only couple of months anyways.

In principle at least they are independent versions. I am pretty sure I asked htuch and he said they are likely to keep it, but I suppose that could always change. At any rate I'd prefer to not block on v3 API support.

tsaarni commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,710 @@
+	// +optional
+	AuthPolicy *AuthorizationPolicy `json:"authPolicy,omitempty"`
+
+        // If FailOpen is true, the client request is forwarded to the upstream service
```
Is this controlling the `status_on_error` on Envoy API ([link](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/http/ext_authz/v3/ext_authz.proto#extensions-filters-http-ext-authz-v3-extauthz))? 
Maybe document the default behavior here to clarify that HTTP 403 Forbidden will be returned by default if authorization server fails to respond.

jpeach replied:
Yup, definitely worth make that clear in the docs.

jpeach commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,710 @@
+  name: testserver
+  namespace: contour-auth
+spec:
+  service:
```
Hah, services became a slice at the last minute and I forgot to update the sample YAML.

jpeach commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,710 @@
+- Most HTTP authorization protocols depend on HTTPS to keep headers and URL query parameters private.
+  Supporting only authorization on HTTPS, Contour reduces the changes of misconfiguring this.
+- HTTPS sessions between Envoy and the authorization server are required.
+  TLS validation can be configured if necessary (should be recommended),
```
Yeh that should be doable eventually.

jpeach commented on design/external-authorization-design.md:
```diff
@@ -0,0 +1,710 @@
+	// +optional
+	Disabled bool `json:"disabled,omitempty"`
+
+	// Context is a set of key/value literal strings that are sent to the
```
Yeh this is context extensions, will make it clear.

glerchundi commented:
Thinking out loud but as this design doc i ntroduces the concept of
ExtensionService, have you considered the idea of using WASM based
extension services instead of just those based on gRPC?

It’s problably premature to take even into consideration but I would like
to put it on the table to see if the new spec definitions can absorb them
in the future.

On Wed, 22 Jul 2020 at 02:23, James Peach <notifications@github.com> wrote:

> Merged #2643 <https://github.com/projectcontour/contour/pull/2643> into
> master.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/projectcontour/contour/pull/2643#event-3573597131>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AARA7FREL5OJQJVWTIFSIYTR4YWQPANCNFSM4OMCA73A>
> .
>

jpeach commented:
> Thinking out loud but as this design doc i ntroduces the concept of ExtensionService, have you considered the idea of using WASM based extension services instead of just those based on gRPC? It’s problably premature to take even into consideration but I would like to put it on the table to see if the new spec definitions can absorb them in the future.

In principle the extensionRef field in the `HTTPProxy` resource could reference something that is not an `ExtensionService` CRD.
